src/test/unistd/test_unistd.rs

```diff
@@ -151,6 +156,177 @@ fn test_getpid_kill() {
+
+    // Creating a directory with an invalid path component should fail.
+    let dirname = CString::new("./nonexistent/foo").unwrap();
+    assert_eq!(
+        linux_api::errno::Errno::result_from_libc_errno(-1, unsafe {
+            libc::mkdir(dirname.as_ptr(), 0)
+        }),
+        Err(linux_api::errno::Errno::ENOENT)
+    );
+}
+
+fn test_mkdirat() {
+    // Create an outer directory first, using AT_FDCWD (equivalent to `mkdir`).
+    let outerdirname = CString::new("./outerdir").unwrap();
+    {
+        let mode = libc::S_IRWXU;
+        linux_api::errno::Errno::result_from_libc_errno(-1, unsafe {
+            libc::mkdirat(libc::AT_FDCWD, outerdirname.as_ptr(), mode)
+        })
+        .unwrap();
+    }
+
+    let dirfd = linux_api::errno::Errno::result_from_libc_errno(-1, unsafe {
+        libc::open(outerdirname.as_ptr(), libc::O_DIRECTORY)
+    })
+    .unwrap();
+    let basename = CString::new("innerdir").unwrap();
+    let mode = libc::S_IRWXU;
+    linux_api::errno::Errno::result_from_libc_errno(-1, unsafe {
+        libc::mkdirat(dirfd, basename.as_ptr(), mode)
+    })
+    .unwrap();
+
+    // Validate that the directory exists with the expected properties.
+    let mut stat: libc::stat = unsafe { std::mem::zeroed() };
+    {
+        let path = CString::new("./outerdir/innerdir").unwrap();
+        linux_api::errno::Errno::result_from_libc_errno(-1, unsafe {
+            libc::stat(path.as_ptr(), &mut stat)
+        })
+        .unwrap();
+    }
+    // Validate that it has the expected permissions.
+    assert_eq!(stat.st_mode, libc::S_IFDIR | mode);
+
+    // Trying to create it again should fail.
+    assert_eq!(
+        linux_api::errno::Errno::result_from_libc_errno(-1, unsafe {
+            libc::mkdirat(dirfd, basename.as_ptr(), 0)
+        }),
+        Err(linux_api::errno::Errno::EEXIST)
+    );
+
+    // Clean up.
+    std::fs::remove_dir("./outerdir/innerdir").unwrap();
+    std::fs::remove_dir("./outerdir").unwrap();
+
+    // Creating a directory with an invalid path component should fail.
+    {
+        let basename = CString::new("nonexistent/foo").unwrap();
+        assert_eq!(
+            linux_api::errno::Errno::result_from_libc_errno(-1, unsafe {
+                libc::mkdirat(dirfd, basename.as_ptr(), 0)
+            }),
+            Err(linux_api::errno::Errno::ENOENT)
+        );
+    }
+
+    // Creating a directory with an invalid file descriptor should fail.
+    assert_eq!(
+        linux_api::errno::Errno::result_from_libc_errno(-1, unsafe {
+            libc::mkdirat(-1, basename.as_ptr(), 0)
+        }),
+        Err(linux_api::errno::Errno::EBADF)
+    );
+
+    // Using a non-directory fd should fail.
+    let tmpfd = {
+        let path = CString::new(".").unwrap();
+        linux_api::errno::Errno::result_from_libc_errno(-1, unsafe {
+            libc::open(path.as_ptr(), libc::O_TMPFILE | libc::O_RDWR, libc::S_IRWXU)
+        })
+        .unwrap()
+    };
```

Comment:
Optional: I think `O_TMPFILE` only works on a few filesystems and not overlayfs, which is used when using a bind mount with docker. This may cause test failures when someone shares the shadow directory from the host to a container since the syscall would fail with `EOPNOTSUPP`, but I don't have docker installed currently so I can't test it so I might be wrong.

For reference, I ran into this before in https://github.com/stevenengler/repatch/issues/1 and I think it applies here to.